nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -181,6 +218,244 @@ impl DataStore {
+                                    vmm_id,
+                                    instance.id(),
+                                    propolis_id,
+                                )
+                            ));
+                        }
+
+                        Ok(())
+                    }
+
+                    (None, Some(dst_propolis_id)) => {
+                        Err(VolumeGetError::CheckoutConditionFailed(
+                            format!(
+                                "InstanceStart {}: instance {} has no propolis id but dst propolis id {}",
+                                vmm_id,
+                                instance.id(),
+                                dst_propolis_id,
+                            )
+                        ))
+                    }
+                }
+            }
+
+            VolumeCheckoutReason::InstanceMigrate { vmm_id, target_vmm_id } => {
+                // Check out this volume to send to destination Propolis to
+                // migrate an Instance. Only take over from the specified source
+                // VMM.
+
+                let Some(instance) = &maybe_instance else {
+                    return Err(VolumeGetError::CheckoutConditionFailed(
+                        format!(
+                            "InstanceMigrate {} {}: instance does not exist",
+                            vmm_id, target_vmm_id
+                        ),
+                    ));
+                };
+
+                let runtime = instance.runtime();
+                match (runtime.propolis_id, runtime.dst_propolis_id) {
+                    (Some(propolis_id), Some(dst_propolis_id)) => {
+                        if propolis_id != *vmm_id || dst_propolis_id != *target_vmm_id {
+                            return Err(VolumeGetError::CheckoutConditionFailed(
+                                format!(
+                                    "InstanceMigrate {} {}: instance {} propolis id mismatches {} {}",
+                                    vmm_id,
+                                    target_vmm_id,
+                                    instance.id(),
+                                    propolis_id,
+                                    dst_propolis_id,
+                                )
+                            ));
+                        }
+
+                        Ok(())
+                    }
+
+                    (None, None) => {
+                        Err(VolumeGetError::CheckoutConditionFailed(
+                            format!(
+                                "InstanceMigrate {} {}: instance {} has no propolis ids",
+                                vmm_id,
+                                target_vmm_id,
+                                instance.id(),
+                            )
+                        ))
+                    }
+
+                    (Some(propolis_id), None) => {
+                        // XXX is this right?
```

Comment:
@gjcolombo what do you think about this? I added it after moving all this to the separate function, and rereading each match case.

Reply:
This is a "there's probably a bug in Nexus" case: we're checking out this volume so that we can use this generation number in a migration destination, but the instance in question has no destination VMM set yet. That's definitely not how the migration saga is supposed to work, so if we hit this case something weird has happened.

That said, allowing the checkout to proceed is probably not a problem vis-à-vis the Crucible generation numbers--in the worst case this generation might go unused, because whoever is asking to migrate will eventually need to win the exclusive right to kick off the migration. I don't think that breaks the other properties we want to ensure with this change.